components/mbedtls_cxx/examples/uart_mutual_auth/main/uart_mutual.cpp

```diff
@@ -0,0 +1,378 @@
+                                  "M95/kVvR1+TwlfS1rSQVUUJXsawUz5iWNroM8YEoxBQ/WA2CGj/w59Oxn446eRbE\n"
+                                  "EPrq8zUCgYEA2g8NC+ZOQMskmdPAX+/wu88CqN1JXOOTTannIjT+XRmvFnI2IcR2\n"
+                                  "Qsz2hSKsNbhJH4yKqmAgncTXFdAadKt5ZjQcLFd9JeG/GoLcTzt6IGgRzl1LKl5Y\n"
+                                  "YMSV5SBI16m9HBXgWnzNgvct67zmtUucFMqSMdMxknzJbG9FYAghL/cCgYEAx7E0\n"
+                                  "rHZHJMZpSmBg5dDO9l3QPrRxxtIQy5o5+5YvQ6fpk4N5F+KtDye3oLjKld2k02sV\n"
+                                  "DJYfr5B9DLqsrRskQM8F/q9D8Az/PLfUpbmZ6WVN0jTAu//HPbZHa0LpH3to+B/S\n"
+                                  "nnkdDWHlZs9TvuQr5YZXuGqnTAvyhkOoEkNN3/sCgYA8DcMZEN9iRtAYsVGc2lbh\n"
+                                  "Uly4JuFqfJ532B/4ssGO4GDw/Jld6V5sfUgzWF43GT7COpGB5KF28dwOfNacZRE1\n"
+                                  "DYrox1uHEEnyQjHsfEPhIugsflMSIxOR6vIhPSfyhSO41WmJYi+zLuHtt4OOUHl2\n"
+                                  "3Gcw46oWXtmWTHq9vN9u9wKBgGiLzt7nwZFwSxmEYdaPvnrfXLIneFW2DtL5eJfN\n"
+                                  "5grOswvmzhQCOcZwbcO4W1+gvbVuH4QKaKZayA1NAjBSwGUpvaK8EZ5wv4QDXlIx\n"
+                                  "XHID9n0x3yHN5HrbnoJ6cmBoFOmqh3MuR1aFRTvRGbAb9xtgfTZwqAu5SYyfiTOe\n"
+                                  "hvvXAoGAfYGO51AR1qfzYQh5nhm4/gFRESUbZevpNDF610GCXNhrcFWH0WgL5rUo\n"
+                                  "bUPvhje/iRpmarv4vB75ax9wTpBp+Wl08YDx4sHlBUAi5JYvieyotfxjI359jAE5\n"
+                                  "X+RPrWjoMOX3QUxCQt5D1gnMi9msFzpsZaEOkeMKf/DTjNNREFA=\n"
+                                  "-----END RSA PRIVATE KEY-----";
+
+/**
+ * Using DTLS the below is set to true.
+ * In that case, we need to receive the entire datagram, not a fragment
+ * This defines a very simple datagram protocol over UART:
+ *  | HEADER (2bytes) | PAYLOAD ...   |
+ *  | dgram_len       | dgram_payload |
+ *
+ * If `use_dgrams` is set to false, we perform TLS on UART stream.
+ * The UART driver is already a stream-like API (using ringbufer), so we simple read and write to UART
+ */
+static const bool use_dgrams = false;
+
+class SecureLink: public Tls {
+public:
+    explicit SecureLink(uart_port_t port, int tx, int rx) : Tls(), uart(port, tx, rx) {}
+    ~SecureLink() = default;
+    int send(const unsigned char *buf, size_t len) override
+    {
+        if (use_dgrams) {
+            // sends a separate dgram header
+            uint16_t header = len;
+            uart_write_bytes(uart.port_, &header, 2);
+        }
+        return uart_write_bytes(uart.port_, buf, len);
+    }
+
+    int recv(unsigned char *buf, size_t len) override
+    {
+        // stream read
+        return uart.recv(buf, len, 0);
+    }
+
+    int recv_tout(unsigned char *buf, size_t len, int timeout) override
+    {
+        // dgram read
+        return uart.recv_dgram(buf, len, timeout);
+    }
+
+    bool open(bool server_not_client)
```

Comment:
I suggest moving this to private and provide two separated public functions:
- connect to open as client
- listen to open as server